implement M3, M4 & M5 spindle control commands (replaces PR #4939 and PR #5113)

Bob-the-Kuhn opened the pull request:
brought config_adv up to date

thinkyhead commented on Marlin/Configuration_adv.h:
```diff
@@ -833,6 +832,106 @@
+ *
+ *  In the pins.h file for your board you'll need to add the following:
+ *   #define SPINDLE_ENABLE_PIN xx     // xx is the digital pin number
+ *   #define SPINDLE_SPEED_PIN yy      // yy is the digital pin number
```
Under the new paradigm we're pre-defining the recommended pins in all the pins files that we can figure out. So if you have some idea of which pins _should_ be used, please add them to the pins files. You should at least add them to the `pins_RAMPS.h` file because this applies to most of the boards we support.

Bob-the-Kuhn replied:
I can do that.  Do you want it done on this PR?

I'm not aware of any semi standard/recommended pins. I'm thinking something like:
//#define SPINDLE_ENABLE_PIN  45   // needs a hardware PWM
//#define SPINDLE_SPEED_PIN    47
//#define CASE_LIGHT_PIN          44   // needs a hardware PWM

thinkyhead commented on Marlin/Configuration_adv.h:
```diff
@@ -833,6 +832,106 @@
+ *  You'll need to select a 3.3V - 5V signal for the ON/OFF function and optionally
+ *  choose a 0-5V PWM pin for the speed control.
+ *
+ *  uncomment the following to enable the ON/OFF function
```
We don't need "uncomment to enable" on every setting. We assume users get that already.

thinkyhead commented on Marlin/Configuration_adv.h:
```diff
@@ -833,6 +832,106 @@
+//#define SPINDLE_ENABLE
+ 
+ //  uncomment the following to enable speed control
+//#define SPINDLE_SPEED  
```
All settings that depend on `SPINDLE_ENABLE` should be inside a block that starts with `#if ENABLED(SPINDLE_ENABLE)`. This allows a putative configuration utility to glean which settings depend on others, and allows users to see this also at a glance.

thinkyhead commented on Marlin/Configuration_adv.h:
```diff
@@ -833,6 +832,106 @@
+ *      TIMER5A   46    Z_STEP                      stepper motor current E PWM
+ *  
+ *  In addition to the above, fans can be assigned to PWM pins.  If you pick a pin that's 
+ *  already assigned to a fan then you'll need to delete the fan or change it's pin assignment.
```
The word should be "its" not "it's."

Bob-the-Kuhn commented:
I think I'm done screwing up for the day.  But there's still daylight out so maybe there's still hope :-(

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -4464,6 +4467,128 @@ inline void gcode_G92() {
+ * for use in the M3 routine.
+ */
+
+  inline void gcode_M4() {
```
Don't name this `gcode_M4` because it's not a formal handler, nor require users to hack code to debug things. It's better if you make a settings flag to call this function instead of `gcode_G3_G4`, and then name the function something like `debug_G4`.

Bob-the-Kuhn replied:
OK - I'll add an option to the M3/M4 command and then document it in the gcodes section of the new wiki.

thinkyhead commented on Marlin/Configuration_adv.h:
```diff
@@ -227,7 +227,6 @@
 // Define a pin to turn case light on/off
 //#define CASE_LIGHT_PIN 4
 //#define CASE_LIGHT_DEFAULT_ON   // Uncomment to set default state to on
-
```
We probably want to preserve this blank line.

Bob-the-Kuhn replied:
They're assigning a pin here rather than in a pins file.  Is that what you want?  

I can include this in the pins file changes you requested for the two spindle pins.

Bob-the-Kuhn commented on Marlin/Configuration_adv.h:
```diff
@@ -833,6 +832,106 @@
+ *
+ *  In the pins.h file for your board you'll need to add the following:
+ *   #define SPINDLE_ENABLE_PIN xx     // xx is the digital pin number
+ *   #define SPINDLE_SPEED_PIN yy      // yy is the digital pin number
```
I can do that.  Do you want it done on this PR?

I'm not aware of any semi standard/recommended pins. I'm thinking something like:
//#define SPINDLE_ENABLE_PIN  45   // needs a hardware PWM
//#define SPINDLE_SPEED_PIN    47
//#define CASE_LIGHT_PIN          44   // needs a hardware PWM

Bob-the-Kuhn commented:
OOPs - spoke too soon.  Scott is reviewing the changes so I have plenty of opportunity to continue screwing up.

Bob-the-Kuhn commented:
All changes requested in the review have been included in this commit.

thinkyhead commented:
After corrections, it's also cool to do a "squash" of commits by running `git rebase -i` and using the "fixup" option in the editor.

Bob-the-Kuhn commented:
The squash/fixup has been completed

msutas commented:
Nice work. But shouldn't there be a pin and an M command for determining the direction of the turn of the spindle. Most of the speed controllers for the spindle have a direction input.

Bob-the-Kuhn commented:
A full implementation would have M3 for clockwise and M4 for counter-clockwise.

I don't have access to a reversible spindle so I can't test the code for it.

We could put a direction pin option (and a reverse flag) in config_adv 

I'd prefer to do that as an enhancement after the current code is merged.  I'd have to dry lab it.

Bob-the-Kuhn commented:
@msutas  - when switching between M3 & M4, does Marlin need to do anything other than change the direction pin?  I'm wondering if Marlin needs to turn off the spindle, wait for it to spin down, change the direction pin and then turn on the spindle.  Someone needs to worry about changing the direction and I don't know if that would be Marlin's job or the spindle controller's job.

thinkyhead commented on Marlin/pins_ELEFU_3.h:
```diff
@@ -147,3 +147,160 @@
+
+
+
+
```
@Bob-the-Kuhn Something went awry with this file, `pins_ELEFU_3.h`.

thinkyhead commented on Marlin/pins_ULTIMAKER_OLD.h:
```diff
@@ -95,3 +95,7 @@
+
+//#define SPINDLE_ENABLE_PIN 44
+//#define SPINDLE_SPEED_PIN  45 // needs a hardware PWM
+//#define CASE_LIGHT_PIN     46 // needs a hardware PWM
```
I think we need to be more authoritative on the pins defined for each board. The `E0_DIR_PIN` is pin 45 in this file, for example.

Also, the idea is that these pins are not commented out, but are explicitly defined for use by the enabled feature. In this case, the feature is `SPINDLE_ENABLE`, enabling any code that refers to the spindle, using these pins.

The same should really apply with the `CASE_LIGHT_PIN` too, but that's another topic.

Bob-the-Kuhn replied:
I can't say that particular pins are normally used for any of the three pins.

It's like a lot of the other functions - way too much variance in the controllers to be able to "dedicate" pins in general.  We might be able to dedicate pins on specific controllers but that would require a lot of work.

I'll gladly do what you think is best.  

I'm thinking our choices are a) back the changes out, or b)leave them as is, or c) open another issue to pursue this question or d) look at each individual board and try to find suitable unused pins.

thinkyhead reviewed (changes requested):
Let's continue to make this better…

Bob-the-Kuhn commented:
Direction control changes were just committed.

I'll start working on the pins_ files tomorrow.

Bob-the-Kuhn commented:
I just did a housekeeping commit.
   1) in config_adv commented out a couple of defines that were keeping the check from passing
   2) restored the pins files (copied over from today's RCBugFix branch)

Bob-the-Kuhn commented:
Modifying the pins_ xxx files is getting to be a big job.

Some of the files have unused pins so I could just add in the new pins.  I'm concerned that the pins I pick may not be physically available on the board.  By chance is there a place where I can look at the hardware manual (pin definitions) for the Marlin supported boards?  I'm hoping to avoid downloading and searching a bunch of schematics.

Adding the new pins to RAMPS is "interesting".  Most pins have multiple names assigned and some names are assigned to almost 10 pins.  I'm surprised that the LCD displays have such a wide variance in the assigned pins.  Is there a hardware reason for this?  Do some of the LCDs come with a single connector that's meant to be plugged into a dedicated header?

thinkyhead commented on Marlin/pinsDebug.h:
```diff
@@ -125,7 +125,10 @@ static bool report_pin_name(int8_t pin, bool &pin_is_analog) {
+  #if PIN_EXISTS(CASE_LIGHT)
+    PIN_SAY(CASE_LIGHT_PIN);
+  #endif
+   #if PIN_EXISTS(CONTROLLERFAN)
```
An extra space crept in here. Take care.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -7721,6 +7836,21 @@ void process_next_command() {
+          break;          // synchronizes with movement commands 
+     #endif
+
+     
```
4 spaces crept in here. You can always regex search `[ \t]+$` and replace with nothing to remove trailing white-space.

thinkyhead commented:
> I'm surprised that the LCD displays have such a wide variance in the assigned pins.

It has to do with how the boards implement the controller/display connectors.

thinkyhead commented:
> the pins I pick may not be physically available on the board. By chance is there a place where I can look at the hardware manual (pin definitions) for the Marlin supported boards?

If you're not sure which pins to assign, you can leave them out for the time-being. You pretty much have to research and download the datasheets for each board to see how the pins are distributed. Generally, for this feature's pins look at the various AUX connectors.

Bob-the-Kuhn commented:
Fixed the space issues.

On Marlin_main.cpp a lot more space issues were fixed than Scott requested.  I selected the whole file, deleted trailing spaces and converted tabs to spaces.

Bob-the-Kuhn commented:
I'd like to proceed with merging this PR as is and create a new PR just for the pins_xxxx.h updates.

I have other commitments so I'll only be doing 1 or 2 boards a day.  There are 60 boards to look at so it'll be 1-2 months before I'll be finished updating the pins_xxxx.h files.

What I'd like to do is create a new PR with only 1 or 2 boards and then update it weekly.

Blue-Marlin commented:
A router is not a 3D-printer.
All 3d-printers have at least one extruder.
A router does not need an extruder.
A extruder stepper driver provides 3 signals: step, dir, enable.
If enable is shared with other stepper drivers the heater pin for this extruder is still not in use.

Converting one extruder is always successful. That should be good enough for a suggestion.
In doubt RepRappers change their pin configurations all day long.

Bob-the-Kuhn commented:
The "fun" part is determining if the pin is physically available.  On some boards the extruder pins are only run to the on board motor drivers.

Bob-the-Kuhn commented:
Just committed the first batch of pins_xxx.h files with M3, M4, M5 spindle support.

I want to have some other eyes look at them so I'm sure I'm on the right track.

This is about 1/3 of the files that need to be looked at and possibly changed.  It includes all the RAMPS based controllers.

I also included support for the case light when possible.

Changed a note in the configuration_adv.h file.

Bob-the-Kuhn commented:
I've found some gaps in my knowledge of how Marlin implements interrupts & PWMs.  There's more useful PWMs available than I thought on the 2560 based controllers.

This will affect the big note in configuration_adv.h and the M43 I report.

It also means that I shouldn't be modifying the pins_xxx.h files for the 644 based controllers until I can either experiment on one or find a trustworthy source.

Looks like it'll be 3-5 weeks before I'll be able to get a 644 controller, do the experiments & update the files.

Bob-the-Kuhn commented:
I've just uploaded a major change to pinsDebug & would like to get comments/feedback before I finish it.

The changes are in the files **pinsDebug.h progmem** & **pinsDebug_list.h** 

It still needs updated/corrected PWM comments.
1.  moved as much as I could into PROGMEM
2. now reports all the names assigned to a pin

The big change was creating an array in FLASH that contained every active pin definition.  That reduced the RAM memory usage considerably but increased the FLASH usage a lot.

Creating the array requires going through the pin list twice.  Rather than having two copies of it in the code I moved the list out to another file and then just did two #includes.

One new item that may cause some concern is the usage of the **LINE** predefined preprocessor macro.  It may not be available if the Arduini IDE goes to a different compiler.

Memory usage changes when enabling PINS_DEBUGGING

  current pinsDebug   FLASH: +10,936   RAM: +742
  proposed system     FLASH: +14,520   RAM: +165

The proposed system saves 577 bytes of RAM at the cost of 3,584 bytes of FLASH.

Bob-the-Kuhn commented:
I'm getting different memory usage numbers today.  I was so surprised by the new numbers for the PROGMEM version of pinsDebug that I downloaded the pinsDebug from RCBugFix, executed both versions and re-ran the compiler 4 times on both versions.

The PROGMEM version takes up less space for BOTH FLASH and RAM.

Memory usage changes when enabling PINS_DEBUGGING flag in configuration_adv.h:
current pinsDebug FLASH: +10,572 RAM: +743
PROGMEM system FLASH:   +8,124  RAM:   +67

The proposed system saves 676 bytes of RAM  and 2,448 bytes of FLASH.

Apparently the executable code buried within the macros does not make it into the PROGMEM compiled image since that code is only executed during compilation.  It more than makes up for the 6K of code added to handle the array.

The current pinsDebug results in 72 copies of`sprintf(buffer, NAME_FORMAT, NAME); SERIAL_ECHO(buffer); pin_is_analog = true; return true;` in the executable image.  With a cost of about 8k of FLASH it looks like each copy costs about 110 bytes.

Why the big change in the numbers?  I must have mistakenly executed one of the earlier versions of PROGMEM when preparing the table.

Bob-the-Kuhn commented:
The PR #5619 change to pins_K8400.h is included in these changes.

Bob-the-Kuhn commented:
This started out as an attempt to squash the latest round of commits but ... 

The good news is that the final (subject to review) files are now in this PR.

Items for discussion:
1. I included defines for CASE_LIGHT_PIN wherever I could.  It's now looking like they're not compatible with the current case light system/code.  Should I back out the comments or should I change the case light code to have an enable flag?
2. The AT90USB boards required a new Arduino extension.  It needs a permanent home.  Should it be a Marlin maintained/owned item or should it remain in my repository or ...?

Bob-the-Kuhn commented:
I've just uploaded a version of the M355 case light code that allows the pin to be defined in either config_adv or the pins_XXX.h file.  If it's defined in the config_adv file then the one in the pins_XXX.h file is ignored.

This keeps it true to the original intent and doesn't force me to take the case light pin definitions out of the pins files.

I've also added some features that make it work pretty much the same way no matter if:
- if the pin is digital only or is a PWM 
- if the "S" parameter or the "P" parameter is used

I've also changed the reporting so that the brightness level is reported if the pin is a PWM.

Bob-the-Kuhn commented:
Just did a commit that will fix the last error being reported by the "checks" routine

Bob-the-Kuhn commented:
I just deleted **pins_99.h** from this PR.  

This file has been removed from RCBugFix sometime after I started this work.  No one wants to see it re-appear.

thinkyhead commented on Marlin/pins_GEN6.h:
```diff
@@ -84,3 +111,15 @@
+  #define SPINDLE_ENABLE_PIN 5      // should have a pull up/down on it
+  #define SPINDLE_SPEED_PIN  16     // MUST BE A HARDWARE PWM
+  #define SPINDLE_DIR_PIN    6
+#endif
```
You should remove the `#if` blocks on these pins and the case light pin below. In Marlin we want to always define the pins for every feature, regardless of whether the feature is enabled or not, so we don't have to have tests like this in every pins file. (It makes them become less manageable.) If there is some conflict between the spindle and some other feature then it should be flagged in `SanityCheck.h`.

thinkyhead commented on Marlin/pins_GEN7_12.h:
```diff
@@ -40,9 +67,14 @@
 // Limit Switches
 //
 #define X_STOP_PIN          7
+#define X_MIN_PIN           7   // added 6 NOV 2016
```
We have either stop pin or min/max pins. We don't have both. The min or max pin is assigned to the stop pin in `pins.h` based on the homing direction.

thinkyhead commented on Marlin/pins_GEN7_14.h:
```diff
@@ -24,8 +24,35 @@
+
 #if !defined(__AVR_ATmega644P__) && !defined(__AVR_ATmega644__) && !defined(__AVR_ATmega1284P__)
-  #error "Oops!  Make sure you have 'Gen7' selected from the 'Tools -> Boards' menu."
+  #error "Oops!  Make sure you have 'SANGUINO' selected from the 'Tools -> Boards' menu."
```
For consistency, `Sanguino` instead of `SANGUINO` here.

thinkyhead commented on Marlin/pins_GEN7_CUSTOM.h:
```diff
@@ -27,8 +27,35 @@
+
 #if !defined(__AVR_ATmega644P__) && !defined(__AVR_ATmega644__) && !defined(__AVR_ATmega1284P__)
-  #error "Oops!  Make sure you have 'Gen7' selected from the 'Tools -> Boards' menu."
+  #error "Oops!  Make sure you have 'SANGUINO' selected from the 'Tools -> Boards' menu."
```
As above.

thinkyhead commented on Marlin/pins_LEAPFROG.h:
```diff
@@ -88,4 +88,4 @@
 #define SOL2_PIN           17
 
-/*  Unused (1) (2) (3) 4 5 6 7 8 9 10 11 12 13 (14) (15) (16) 17 (18) (19) (20) (21) (22) (23) 24 (25) (26) (27) 28 (29) (30) (31)  */
+/*  Unused (1) (2) (3) 4 5 6 7 8 9 10 11 12 13 (14) (15) (16) 17 (18) (19) (20) (21) (22) (23) 24 (25) (26) (27) 28 (29) (30) (31)  */
```
If not needed, kill this change. One fewer file in the commit log.

thinkyhead commented on Marlin/pins_MEGATRONICS_3.h:
```diff
@@ -166,3 +168,29 @@
+  #define SPINDLE_SPEED_PIN      4
+  #define SPINDLE_ENABLE_PIN    17  // should have a pullup resistor on this pin
+  #define SPINDLE_DIR_PIN        5
+#endif
```
This block is ok. No objections.

thinkyhead commented on Marlin/pins_MIGHTYBOARD_REVE.h:
```diff
@@ -220,3 +231,19 @@
+  #define SPINDLE_ENABLE_PIN 66 // K4   should be pulled up
+  #define SPINDLE_SPEED_PIN   8 // H5   needs a hardware PWM
+  #define SPINDLE_DIR_PIN    67 // K5
+#endif
```
As above, define these pins unconditionally.

Mightyboard is still not working, and might never, but we'll see! Thank you for the B mosfet correction.

thinkyhead commented on Marlin/pins_MINITRONICS.h:
```diff
@@ -123,3 +130,22 @@
+  #define SPINDLE_ENABLE_PIN 52 // using A6 because it already has a pull up on it
+  #define SPINDLE_SPEED_PIN   3 // WARNING - LED & resistor pull up to +12/+24V stepper voltage
+  #define SPINDLE_DIR_PIN    53
+#endif
```
Here, the dummy assignments are kind of a hack, though I get the reasoning. It's a lot more work to make Marlin support `#define EXTRUDERS 0`. But maybe that would be good to support in the long run.

thinkyhead commented on Marlin/pins_MKS_BASE.h:
```diff
@@ -36,4 +36,31 @@
+  #undef  Z_MAX_PIN
+  #define SPINDLE_SPEED_PIN      2
+  #define SPINDLE_ENABLE_PIN    15  // should have a pullup resistor on this pin
+  #define SPINDLE_DIR_PIN       19
```
Here I'd say it's better to have these pins defined ahead of the `#include` line, then just make them overrideable in the `pins_RAMPS.h` file instead with the usual `#ifndef ... #define ... #endif`.

thinkyhead commented on Marlin/pins_MKS_BASE.h:
```diff
@@ -36,4 +36,31 @@
+    #undef  X_MAX_PIN
+    #define CASE_LIGHT_PIN         2
+  #endif
+#endif
```
Really…? Can you imagine if every feature went this route? Surely we can do better. If there are incompatibilities between settings, I would rather we flag them in `SanityCheck.h` than simply undefine pins to prevent their use. This relies on assumptions about the way the code will react to a missing x-max-pin, which is unspecified.

thinkyhead commented on Marlin/pins_PRINTRBOARD.h:
```diff
@@ -21,18 +21,156 @@
 
 #ifndef __AVR_AT90USB1286__
-  #error "Oops!  Make sure you have 'Teensy++ 2.0' selected from the 'Tools -> Boards' menu."
+  #error "Oops!  Make sure you have 'Teensy++ 2.0', "AT90USB1286_TEENSYPP", "AT90USB1286_STANDARD" or "Printrboard" selected from the 'Tools -> Boards' menu."
```
Single quotes will be better here.

thinkyhead commented on Marlin/pins_PRINTRBOARD.h:
```diff
@@ -21,18 +21,156 @@
+#elif defined(ARDUINO_AT90USB1286_STANDARD)
+  #define USE_TEENSYPP_PIN_MAP false // using STANDARD pin map
+#else
+  #define USE_TEENSYPP_PIN_MAP true     // default to TEENSYPP pin map
```
For the pin mapping stuff, first, please make a separate PR from this one and remove the pin mapping-related changes from this PR. Second, let's not include both pin mappings in a single file. Use these conditions to choose which pins file to include instead, then put the pins definitions into new files `pins_PRINTRBOARD-Teensy.h` and `pins_PRINTRBOARD-AT90USB.h`.

thinkyhead commented:
Please review my latest comments and make the requested changes. Sorry my review has taken so long. I'm in the midst of a long-distance relocation.

Bob-the-Kuhn commented:
I haven't forgotten about this PR.  It'll be a few more days before I'll have a big enough chunck of time to work on it.

Bob-the-Kuhn commented:
Just did a squash so I know for sure what is on this PR.  

Looks like some of the "remove M355 case light changes" have disappeared.  I'll have to remedy this.

I'll let you know when this PR is ready for review again.

Bob-the-Kuhn commented:
Strange - "INVERT_CASE_LIGHT" is still causing problems but not as many as before ???

Will try doing a squash which'll re-run the checks.

Bob-the-Kuhn commented:
All inputs from the review have been implemented.

This PR is ready for review.

Bob-the-Kuhn commented:
@thinkyhead - you mentioned that Mightyboard isn't working yet.  I'd be glad to help.

I cut my engineering teeth on microprocessor based boards so many years ago.  That was when the 8085 was big news.

I've got an OK logic analyzer and, if I can ever get the third party software running, a low bandwidth 'scope.

Bob-the-Kuhn commented:
@thinkyhead 

There was a comment about wanting to include some new pin names before merging this PR.  Are they available now?

Bob-the-Kuhn committed:
all code changes are in Marlin_main.cpp

none of the existing code was changed

section added to config_adv to enable/disable the new commands & to explain the hardware changes needed to properly connect to the spindle controller

There's lots of competition for the PWM pins so added a section in sanity.h to check that the selected pin doesn't conflict with other selections.

==============================================================

added spindle pin definitions and case light definitions to all the pins_XXXX.h files that would support these functions

===============================================================

added M355 code cleanup/improvements

===============================================================

removed M355 code cleanup/improvements - they are being done on a separate PR

===============================================================

multiple rounds of cleanup

===============================================================

Moved changes for pins_5DPRINT.h, pins_PRINTRBOARD.h & pins_PRINTRBOARD_REVF.h to a new PR.  These boards all use a pin that is undefined in the Teensyduino extension to Arduino.  The fix is pretty major.

===============================================================

make INVERT_CASE_LIGHT always defined

An unintended consequence of adding the CASE_LIGHT_PIN definition to the
pins_XXX.h files is that the compile will fail for all boards that have
the pin defined.

The short term fix is to change Marlin_main.cpp so that INVERT_CASE_LIGHT
always defined.

This is consistent with the long term fix being pursued in PR 5685.

This'll be modified by the M355 PR when it gets implemented.

==============================================================

bring files up to date & add laser tags/names

Bob-the-Kuhn commented:
Bringing everything up to date.

Added laser text to the Configuration_adv.h files. 

@thinkyhead - will this be merged before we go to 1.1.0?

Bob-the-Kuhn committed:
Merge branch 'RCBugFix' into spindle-2-NOV

svogl commented:
@Bob-the-Kuhn - thanks for the pointers, I integrated you PR into my Marlin fork at https://github.com/voxel-dot-at/Marlin and it works well so far, I've dropped a note to the MakeBlock forum as well:

http://forum.makeblock.com/t/laserbot-updated-firmware-to-marlin-1-1-0-rc8/7485

I did change the power_up/_down times to work with milliseconds instead of seconds to get fine-grain timing control of laser power-up (I saw 1-2mm not being cut after turning on the PWM [the laser is powered via the PWM... ahemm...] ). 
I don't know about real spindles, but maybe you want to switch ms as well.

Bob-the-Kuhn commented:
Thanks for the feedback.  I'll change it to milliseconds.

Bob-the-Kuhn committed:
convert delay to milliseconds, correct names, enable only one equation

Bob-the-Kuhn commented:
When I went to resolve the conflicts there were about 20 more than listed here.  Some files were on this PR, some weren't.

Rather than fight GIT I've replaced this PR with PR #6221
